docs/specs/user-model.md

```diff
@@ -312,3 +294,164 @@ Once used, a recovery code is invalidated.
 
 The codes are cryptographically secure random 10-letter string in Crockford's
 Base32 alphabet.
+
+## Claims
+
+Claims are computed information about the user. The computation is controlled by [claims mapping](#claims-mapping).
+
+## Claims Mapping
+
+A user has their own claims mapping. The claims mapping of a user can be updated via Admin API.
+
+> TODO: We may allow the user change some claims mapping, for example, we may allow the user to change which identity to map `email`.
+> However, it is unlikely that the developer would expect the user to change `custom_attributes` mapping.
+
+### Claims Mapping Example
+
+The following example illustrates the full capability of claims mapping.
+
+Given the following claims mapping
+
+```yaml
+mapping:
+- kind: "identity"
+  identity_id: "identity-a"
+  name_pointer: "#/email"
+  value_pointer: "#/email"
+- kind: "custom_attributes"
+  name_pointer: "#/zoneinfo"
+  value_pointer: "#/profile/preferred_timezone"
+- kind: "custom_attributes"
+  name_pointer: "#/picture"
+  value_pointer: "#/profile/profile_image_url"
+- kind: "custom_attributes"
+  name_pointer: "#/app:rbac"
+  value_pointer: "#/rbac"
+```
+
+And the claims of `identity-a`
+
+```json
+{
+  "email": "user@example.com"
+}
+```
+
+And the custom attributes
+
+```json
+{
+  "profile": {
+    "preferred_timezone": "Asia/Hong_Kong",
+    "profile_image_url": "https://cdn.example.com/u/user-a.jpg"
+  },
+  "rbac": [
+    "product:list",
+    "product:get",
+    "product:delete"
+  ]
+}
+```
+
+The computed claims of the user is
+
+```json
+{
+  "email": "user@example.com",
+  "zoneinfo": "Asia/Hong_Kong",
+  "picture": "https://cdn.example.com/u/user-a.jpg",
+  "app:rbac": [
+    "product:list",
+    "product:get",
+    "product:delete"
+  ]
+}
+```
+
+### Automatic addition of claims mapping
+
+When a new identity is being added, for each Standard Claims of the identity, if that claim is not mapped, then a new mapping is created from the claim.
+
+For example, if a new user signs up with Google, they will have the following claim mappings
+
+```
+mapping:
+- kind: "identity"
+  identity_id: "google"
+  name_pointer: "#/email"
+  value_pointer: "#/email"
+```
+
+So the user will have the claim `email` immediately available in their claims.
+
+### Dangling reference in claims mapping
+
+If a deletion of identity will result in dangling reference in the claims mapping, the deletion is disallowed.
```

Comment:
Would it result in the following unexpected behaviour?
- User sign up with email
- User login with Google
- Google's email claim map to the user's email claim
- At User Setting page, the user can never remove the Google login, but he don't understand why... and could never delete his google login because the developer also don't understand why.

Reply:
or the mapping is not "generate the claims everytime claims was queried", but more like "automatically fill in according to claims change"? (random idea.... )

Reply:
or it mapping need to have some way to handle situation like the above, either:

1. If Google login was removed, the next default is the email identities
2. If Google login was removed, the email remains in the claims but unverified.

Both way seems complicated?

Reply:
> Would it result in the following unexpected behaviour?

The error message can explain why deletion is impossible and offer ways to resolve the error

1. Do not delete the identity at all. The information from this identity is part of the public user info (i.e. claims). Deleting the identity WILL change the public user info.
2. Remove the mapping first. It is equivalent to making the user to aware such action is affecting the claims.

---

> or the mapping is not "generate the claims everytime claims was queried", but more like "automatically fill in according to claims change"? (random idea.... )

Not sure what you meant. Can you explain in more details?

---

> If Google login was removed, the next default is the email identities

This seems surprising to me.

---

> If Google login was removed, the email remains in the claims but unverified.

This is actually an alternate design choice. So in this design, we do not have `kind: identity` claim mapping. Every claim mapping entry is just `kind: custom_attributes`. But specifically for the claim `email` and the claim `phone_number`, we will use available information to provide `email_verified` and `phone_number_verified`.

Personally I think the original design (`kind: identity`) offers a stronger guarantee. The claim `email` is very highly to be verified (at least from Authgear's point of view).

Reply:
> The error message can explain why deletion is impossible and offer ways to resolve the error

Can you try to propose the most user friendlys flow and error message, when users delete their Google Login but got rejected? And how do we visualize the mapping of claims in Setting page for layman users?

> Not sure what you meant. Can you explain in more details?
I meant it is not a mapping that's generative / delcarative, but runs the mapping only once when a value change, effectively that will make it like if Google Login was removed, the email remains in the claims.